### PR TITLE
[TOPI][RELAY][ONNX] Scatter ND

### PR DESCRIPTION
**Pull request by mbrookhart**
This PR refactors the relay scatter_nd op in two ways:
1) scatter_nd now takes in a data tensor, and updates are scattered based on indices. The previous API assumed zero-initialization for the data.
2) scatter_nd now has a "mode" argument to determine how to update the input data, either "add" or "update". This is to support the use cases as a gradient of gather_nd and to support the ONNX and pytorch APIs. TF also supports "max", "min", and "sub", which are not yet supported here.

This also adds ScatterND to the ONNX importer.

cc @tkonolige @altanh @jwfromm @masahi

### CODE REVIEW

File: python/tvm/topi/cuda/scatter.py [hunk 1]
```diff
@@ -723,7 +723,7 @@ def update_func(dst_ptr, dst_index, update):
     return out
 
 
-def scatter_nd(data, indices, shape):
+def scatter_nd(data, indices, updates, mode):
     """Scatter elements from a n-dimension array.
 
     Given data with shape (Y_0, ..., Y_{K-1}, X_M, ..., X_{N-1}), indices with shape
```
Comment (tkonolige) [inline python/tvm/topi/cuda/scatter.py:729]: Can you update this?

File: python/tvm/topi/x86/scatter.py [hunk 2]
```diff
@@ -46,62 +46,69 @@ def scatter_nd(data, indices, shape):
     indices : tvm.te.Tensor
         The indices of the values to extract.
 
-    shape : Sequence[int]
-        The output shape. This must be specified because it cannot be inferred.
+    updates : tvm.te.Tensor
+        The updates to apply at the Indices
+
+    mode : string
+        The update mode for the algorith, either "update" or "add"
```
Comment (tkonolige) [inline python/tvm/topi/x86/scatter.py:53]:
```suggestion
        The update mode for the algorithm, either "update" or "add"
```

File: python/tvm/topi/cuda/scatter.py [hunk 3]
```diff
@@ -789,38 +790,41 @@ def gen_ir(data_ptr, indices_ptr, out_ptr):
...
                     index += offset * indices[i + l * fused_indices_dimension]
-                    offset *= shape[l]
-                out[index] += data[i * fused_data_dimension + j]
+                    offset *= data_ptr.shape[l]
+                if mode == "update":
+                    out[index] = updates[i * fused_updates_dimension + j]
+                elif mode == "add":
+                    out[index] += updates[i * fused_updates_dimension + j]
+                else:
+                    raise NotImplementedError("scatter_nd mode not supported:", mode)
```
Comment (tkonolige) [inline python/tvm/topi/cuda/scatter.py:819]: Please add the supported modes to the error message.

File: python/tvm/topi/scatter.py [hunk 4]
```diff
@@ -248,29 +248,31 @@ def scatter_nd(data, indices, shape):
     indices : tvm.te.Tensor
         The indices of the values to extract.
 
-    shape : Sequence[int]
-        The output shape. This must be specified because it cannot be inferred.
+    updates : tvm.te.Tensor
+        The updates to apply at the Indices
+
+    mode : string
+        The update mode for the algorith, either "update" or "add"
```
Comment (tkonolige) [inline python/tvm/topi/scatter.py:255]: What do the different modes do?

**Review by jwfromm — APPROVED** [review]
Very nice, LGTM.

**Review by altanh — APPROVED** [review]
LGTM 👌

**Comment by jcf94** [comment]
Thanks! @mbrookhart @tkonolige @jwfromm @altanh

**Pull request merged**